Core/include/Acts/Geometry/BlueprintNode.hpp

```diff
@@ -0,0 +1,295 @@
+          callback = {});
+
+  /// Convenience method for creating a @ref Acts::LayerBlueprintNode.
+  /// @param name The name of the layer node.
+  /// @param callback An optional callback that receives the node as an argument
+  LayerBlueprintNode& addLayer(
+      const std::string& name,
+      const std::function<void(LayerBlueprintNode& layer)>& callback = {});
+
+  /// @}
+
+  /// Register a @p child to this node.
+  /// @warning This method throws if adding the child would create a
+  ///          cycle in the blueprint tree!
+  /// @param child The child node to add
+  /// @return A reference this node (not the child!)
+  BlueprintNode& addChild(std::shared_ptr<BlueprintNode> child);
+
+  /// A range-like object that allows range based for loops and index access.
+  /// This type's iterators and accessors return mutable references when
+  /// dereferenced.
+  using MutableChildRange =
+      detail::TransformRange<detail::Dereference,
+                             std::vector<std::shared_ptr<BlueprintNode>>>;
+
+  /// A range-like object that allows range based for loops and index access.
+  /// This type's iterators and accessors return const references when
+  /// dereferenced.
+  using ChildRange =
+      detail::TransformRange<detail::ConstDereference,
+                             const std::vector<std::shared_ptr<BlueprintNode>>>;
+
+  /// Return a @ref MutableChildRange to the children of this node.
+  /// @return A range-like object to the children
+  MutableChildRange children();
+
+  /// Return a @ref ChildRange to the children of this node.
+  /// @return A range-like object to the children
+  ChildRange children() const;
+
+  /// Remove all children from this node
+  void clearChildren();
+
+  /// Return the depth of this node in the blueprint tree. A depth of zero means
+  /// this node does not have a parent.
+  /// @return The depth of this node
+  std::size_t depth() const;
+
+  /// Print the node tree starting from this node to graphviz format
+  /// @param os The stream to print to
+  void graphviz(std::ostream& os) const;
+
+  /// Method that writes a representatiohn of **this node only** to graphviz.
+  /// This should generally not be called on its own, but through the @ref
+  /// BlueprintNode::graphviz method.
+  /// @param os The stream to print to
+  virtual void addToGraphviz(std::ostream& os) const;
+
+  /// Print a representation of this node to the stream
+  /// @param os The stream to print to
+  /// @param node The node to print
+  /// @return The output stream
+  friend std::ostream& operator<<(std::ostream& os, const BlueprintNode& node) {
+    node.toStream(os);
+    return os;
+  }
+
+ protected:
+  /// Virtual method to determine stream representation.
+  /// @note This method is called by the stream operator.
+  virtual void toStream(std::ostream& os) const;
+
+  /// Set the depth to @p depth and update children recursively
+  void setDepth(std::size_t depth);
```

Comment:
True. This is a tradeoff vs having to book-keep the parent. I can change it, although it would be an internal refactor, since only the depth value itself is exposed, whether it's pre-calculated or calculated on the fly. 

What do you think?

Reply:
No I think the implementation is fine. I just wondered if it would be beneficial to be able to cross the tree in both directions.